File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int res = 0;
        for(int i = 0; i < grid.size(); ++i) {
            for(int j = 0; j < grid[0].size(); ++j) {
                if(grid[i][j] == '1') {
                    ++res;
                    dfs(grid, i, j);
                } else {
                    continue;
                }
            }
        }
        return res;
    }
    void dfs(vector<vector<char>>& grid, int r, int c) {
        if(r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size() || grid[r][c] == '0') {
            return;
        }
        grid[r][c] = '0';
        dfs(grid, r + 1, c);
        dfs(grid, r - 1, c);
        dfs(grid, r, c - 1);
        dfs(grid, r, c + 1);
    }
};

File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:

        res = []
        stack = []

        def dfs(opn, cls):
            
            if opn == n and cls == n:
                res.append("".join(stack))
            elif opn > n:
                return
            else:
                if cls > opn:
                    return
                elif cls == opn:
                    stack.append('(')
                    dfs(opn + 1, cls)
                    stack.pop()
                    return
                else: # cls < opn
                    stack.append('(') # add opn bracket
                    dfs(opn + 1, cls)
                    stack.pop()
                    
                    stack.append(')') # add cls bracket
                    dfs(opn, cls + 1)
                    stack.pop()

        dfs(0, 0)

        return res
